pkg/ccl/sqlproxyccl/proxy_handler_test.go

```diff
@@ -590,90 +480,341 @@ func TestDenylistUpdate(t *testing.T) {
 	})
 	defer func() { _ = os.Remove(denyList.Name()) }()
 
 	url := fmt.Sprintf("postgres://root:admin@%s/defaultdb_29?sslmode=require&options=--cluster=dim-dog-28", addr)
-	conn, err := pgx.Connect(context.Background(), url)
-	require.NoError(t, err)
-	defer func() {
-		require.NoError(t, conn.Close(ctx))
+	te.TestConnect(ctx, t, url, func(conn *pgx.Conn) {
+		require.Eventuallyf(
+			t,
+			func() bool {
+				_, err = conn.Exec(context.Background(), "SELECT 1")
+				return err != nil
+			},
+			time.Second, 5*time.Millisecond,
+			"Expected the connection to eventually fail",
+		)
+		require.EqualError(t, err, "unexpected EOF")
 		require.Equal(t, int64(1), s.metrics.ExpiredClientConnCount.Count())
-	}()
-
-	require.Eventuallyf(
-		t,
-		func() bool {
-			_, err = conn.Exec(context.Background(), "SELECT 1")
-			return err != nil
-		},
-		time.Second, 5*time.Millisecond,
-		"Expected the connection to eventually fail",
-	)
-	require.EqualError(t, err, "unexpected EOF")
+	})
 }
 
 func TestProxyAgainstSecureCRDBWithIdleTimeout(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
 	ctx := context.Background()
+	te := newTester()
+	defer te.Close()
 
 	sql, _, _ := serverutils.StartServer(t, base.TestServerArgs{Insecure: false})
 	sql.(*server.TestServer).PGServer().TestingSetTrustClientProvidedRemoteAddr(true)
+	defer sql.Stopper().Stop(ctx)
 
 	outgoingTLSConfig, err := sql.RPCContext().GetClientTLSConfig()
 	require.NoError(t, err)
 	proxyOutgoingTLSConfig := outgoingTLSConfig.Clone()
 	proxyOutgoingTLSConfig.InsecureSkipVerify = true
 
 	idleTimeout, _ := time.ParseDuration("0.5s")
-	var connSuccess bool
-	frontendAdmit := frontendAdmit
-	defer hookFrontendAdmit(func(conn net.Conn, incomingTLSConfig *tls.Config) (net.Conn, *pgproto3.StartupMessage, error) {
-		return frontendAdmit(conn, incomingTLSConfig)
-	})()
-	backendDial := backendDial
-	defer hookBackendDial(func(msg *pgproto3.StartupMessage, outgoingAddress string, tlsConfig *tls.Config) (net.Conn, error) {
-		return backendDial(msg, sql.ServingSQLAddr(), proxyOutgoingTLSConfig)
-	})()
-	originalAuthenticate := authenticate
-	defer hookAuthenticate(func(clientConn, crdbConn net.Conn) error {
-		err := originalAuthenticate(clientConn, crdbConn)
-		connSuccess = err == nil
-		return err
+	originalBackendDial := backendDial
+	defer testutils.TestingHook(&backendDial, func(
+		msg *pgproto3.StartupMessage, outgoingAddress string, tlsConfig *tls.Config,
+	) (net.Conn, error) {
+		return originalBackendDial(msg, sql.ServingSQLAddr(), proxyOutgoingTLSConfig)
 	})()
 
-	defer sql.Stopper().Stop(ctx)
-
 	s, addr := newSecureProxyServer(ctx, t, sql.Stopper(), &ProxyOptions{IdleTimeout: idleTimeout})
 
 	url := fmt.Sprintf("postgres://root:admin@%s/?sslmode=require&options=--cluster=dim-dog-28", addr)
+	te.TestConnect(ctx, t, url, func(conn *pgx.Conn) {
+		require.Equal(t, int64(1), s.metrics.CurConnCount.Value())
+		require.Equal(t, int64(1), s.metrics.SuccessfulConnCount.Count())
```

Comment:
You need to run some stress testing to check for races. The successful connection count is increased *after* the server sends back a message confirming the successful password authentication. So you don't have guarantee that you won't do the check before the increase. Previously this check was deferred so the stopper's stop was called before it.